docs: add ops tasks procedure (#181)
iearn-finance/yearn-vaults: docs/OPERATIONS.md

# Deploy a Strategy / Vault for V2

NOTE: This repo is encouraged to create multiple scripts for governance and dev multisig execution of complex transactions
https://github.com/iearn-finance/chief-multisig-officer

**IMPORTANT**: you should **NOT** create a new release with a test vault

## Before deploying
- [ ] Coordinate with Core Dev strategist for getting a review in board https://github.com/orgs/iearn-finance/projects/5
- [ ] Peer review completed by:
    - yearn.rocks: at least 2 strategists

>     - Experimental: at least 1 strategist + 1 core dev in board 

Author: ```suggestion
    - Experimental: at least 1 strategist + 1 core dev in the Code Review board 
```

- [ ] Check if want token has a deploy vault already (>=v0.3.0) and coordinate to use that first if possible.

> - [ ] (only if a new vault is needed) Deploy using the registry.

Reviewer: Question: Given the room to make errors in parameters and the need for us to verify that those are correct in any case, should the Strategist really be doing the vault deployment, or is it better this is handled by dedicated people?

Author: It's a great question. I've added this to my list of tasks to revisit when we have all the process documented. I'd like to have the current process, with a nice amount of details and then we can improve it. What do you think?

    - `registry.newExperimentalVault(want, you, multisig, treasury, "", "")`
        - args: token, gov, guardian, rewards, name, symbol
        - multisig (strategists) = '0x16388463d60FFE0661Cf7F1f31a7D658aC790ff7'
        - gov for testing = '0x846e211e8ba920b353fb717631c015cf04061cc9' (dev.ychad.eth) 
        - treasury = '0x93A62dA5a14C80f265DAbC077fCEE437B1a0Efde'
        (treasury.ychad.eth)

>     - [ ] Coordinate with core devs proper cap limit for new vault and other settings

Reviewer: Are caps arbitrary? If so, maybe mention what the criteria to decide on limits are. 

If possible, would be even better to have some kind of table for what types of limits we set, and how to progress to the next stage.

Author: Caps are more or less arbitrary. Nevertheless, we need this you are suggesting. Will add. Nice catch.

Author: Caps added.

>     - [ ] Check new vault has ABI setup on etherscan (If Verification with vyper and proxy is not working that is ideal)

Author: ```suggestion
    - [ ] Check new vault has ABI setup on Etherscan (until verification using Vyper and proxy works on Etherscan)
```

    - [ ] Set a deposit limit to the vault
    vault.setDepositLimit(50_000 * 1e18)
    $50k USD converted to your want. Example above is 50k DAI
- [ ] Deploy strategy with settings and upload code to Etherscan for verification
- [ ] Tag github review issue https://github.com/orgs/iearn-finance/projects/5 with deployed version. Add mainnet addresses

## After deploying strategy
- [ ] Add strategy to vault
    - `vault.addStrategy(strategy, debt_ratio, rate_limit, 1000)`
    - debt_ratio should be 9800 if first strategy on vault
    - rate_limit is 0 unless there is reason for it to be different

- [ ] Set keeper
    - `strategy.setKeeper(keep3r_manager)`
    - keep3r_manager = '0x13dAda6157Fee283723c0254F43FF1FdADe4EEd6'
- [ ] Set rewards
    - `strategy.setRewards(address)`
    Read below if you want to use the sharer contract
- [ ] Set management fee to 0
    - `vault.setManagementFee(0)`
- [ ] Set governance
    - `vault.setGovernance(multisig)`
    - multisig = '0x846e211e8ba920b353fb717631c015cf04061cc9'
    (dev.ychad.eth)
    - Governance needs to be accepted before it is in place. After you set this you will still have power until part 2.

> - [ ] Run tests against "live" vault and strategy in mainnet-fork

Reviewer: Can we describe a set of standard tests here that we want to make sure work as they should for every vault/strategy?

Reviewer: I see some tests are described further below, perhaps reference them in this section?

Author: Tests described and also op tests added to the brownie-mix :)

- Set management fee to 0
    - `vault.setManagementFee(0)`
- Set governance
    - `vault.setGovernance(multisig)`
    - multisig = '0x846e211e8ba920b353fb717631c015cf04061cc9'
    (dev.ychad.eth)
    - Governance needs to be accepted before it is in place. After you set this you will still have control over the strategy.
- Run tests against "live" vault and strategy in mainnet-fork
    - Harvest
    - Profitable harvest
    - Revoke strategy and check that funds return to the vault
    - Increase/decrease debt + harvest, and check that the strategy is working well
    - Migration
    - Check that tokens in the strategy cannot be sweeped by dust collection

   Example: https://github.com/Macarse/yhegic/tree/master/tests/development
- Tag vault as "experimental" in v2.registry.ychad.eth
    - `registry.tagVault(vaultAddr, "https://meta.yearn.network/vaults/${vaultAddr}/vault.json")`

### Sharer contract (optional)
Sharer is a contract for distributing strategist rewards. For boarding school graduates suggested is 34% to strategist_ms and 66% to strategist.

Repo: https://github.com/Grandthrax/Sharer

- Setup rewards for your strategy by calling `sharer.addContributors`

> - If you forked someone elses strategy then cut them in

Reviewer: *someone else's

    - Talk to Facu
- [ ] Monitor harvest and tend triggers for first few days. Call harvest/tend manually

## Setting up Keep3r
- [ ] Adjust trigger variables until they are correct
    - `strategy.setProfitFactor()`
    - `strategy.setDebtThreshold()`
    - `strategy.setMaxReportDelay()`
- [ ] Add strategy to the keep3r job
    - `keep3r_manager.addStrategy(strategy, 1_500_000, 1_500_000)`
    - Tell Carlos harvest and tend gas usage. For instance 1.5m.

## Scaling Up / Moving to Endorse
- [ ] In additon to the 2 strategists, a Core Developer has to review the strategy before going into production.
- [ ] Increase limits
- [ ] Add to experimental tab on yearn.finance
- [ ] Set management fee to production level
    - `vault.setManagementFee(200)`
- [ ] Set governance to ychad.eth
- [ ] yearn governance now must accept governance and endorse
    - `strategy.acceptGovernance()` should be run by ychad.eth
    - `registry.endorseVault(strategy)` Needs ychad.eth governace accepted for this to work
    - **Order is important. Will fail if order is wrong**
- [ ] Now you are on main yearn page!

## Limits per Stage

> These are the stantdard deposit limits per stage. This can be increased or decreased on a case by case basis.

Reviewer: *standard

> These are the stantdard deposit limits per stage. This can be increased or decreased on a case by case basis.

Reviewer: *They can be adjusted on a case by case basis.

| Stage | Limit |
| ---- | --- |

> | Yearn.rocks / Experimental | $500.000 |

Reviewer: $500.000 → $500,000 (or $500K)
$10.000.000 → $10,000,000 (or $10M)

- Check latest version in `v2.registry.ychad.eth` against the planned new release vault to be sure its an updated version.
- Deploy vault for production using the new version (this should not be an experimental vault since it will be endorsed with this process).
- Set governance to `ychad.eth`:

  ```python
  vault.setGovernance(0xfeb4acf3df3cdea7399794d0869ef76a6efaff52)
  ```

- Let multisig accept governance:

  ```python
  vault.acceptGovernance()
  ```

- Let governance create a new release on `v2.registry.ychad.eth`:

  ```python
  registry.newRelease(vault)
  ```

**Note**: Last two steps may need to be done in different transactions since it sometimes can fail in practice using multisig from Gnosis.

## Before deployment

- Coordinate with Core Dev strategist for getting a review on [board](https://github.com/orgs/iearn-finance/projects/5).
- Complete peer review by at least 2 strategists.
- Check if `want` token has a deployed vault already (>=v0.3.0) and coordinate to use that first if possible.
- If a new vault is needed, deploy it using the registry:

  ```python
  registry.newExperimentalVault(want, you, multisig, treasury, "", "")
  ```

>   - args: token, gov, guardian, rewards, name, symbol

Reviewer 2: What are those?

Author: >  - args: token, gov, guardian, rewards, name, symbol
Is the description of the arguments for this function: registry.newExperimentalVault()

Author: Rewrote it in a more script-like style

  - `multisig` (strategists) = '0x16388463d60FFE0661Cf7F1f31a7D658aC790ff7'

>   - gov for testing = '0x846e211e8ba920b353fb717631c015cf04061cc9' (dev.ychad.eth)

Reviewer 2: And where is this one coming from?

Author: That's the data to replace the variable names here:

registry.newExperimentalVault(want, you, multisig, treasury, "", "")

Author: Rewrote it in a more script-like style

    (`dev.ychad.eth`).
  - Governance needs to be accepted before it is in place. After you set this you will still have control over the strategy.

- Run tests against "live" vault and strategy in mainnet-fork:

  - Harvest.
  - Profitable harvest.
  - Revoke strategy and check that funds return to the vault.
  - Increase/decrease debt + harvest, and check that the strategy is working well.
  - Migration.
  - Check that tokens in the strategy cannot be sweeped by dust collection.

  Example: Hegic strat [repo](https://github.com/Macarse/yhegic/tree/master/tests/development).

- Tag vault as "experimental" in v2.registry.ychad.eth
  - `registry.tagVault(vaultAddr, "https://meta.yearn.network/vaults/${vaultAddr}/vault.json")`

### Sharer contract (optional)

"Sharer" is a contract for distributing/splitting strategist rewards. For boarding school graduates suggested split is 34% to strategist multisig and 66% to strategist – [repo](https://github.com/Grandthrax/Sharer).

- Setup rewards for your strategy by calling `sharer.addContributors`.
- Include devs if you forked someone else's strategy.
- Be sure to reward people who helped you.

### Example Script

```python

> # TBD Fill CMO example script to fill

Reviewer 2: `TODO:`

Reviewer 2: Will be added in next PR.

```

## The Manual Phase

- [ ] Deposit some `want` tokens into the vault.
- [ ] Do first `harvest` and make sure it worked correctly.

  ```python
  strategy.harvest()
  ```

- [ ] Publish on yearn.rocks:
  - Governance dev multisig **_must_** call `vault.acceptGovernance()` first.

>   - Talk to Facu.

Reviewer 2: How do I contact him?

Author: Removed. Added instructions to create a PR.

- [ ] Monitor `harvest` and `tend` triggers for first few days. Call `harvest`/`tend` manually.

## Setting up Keep3r

- [ ] Adjust trigger variables until they are correct:
  - `strategy.setProfitFactor()`
  - `strategy.setDebtThreshold()`
  - `strategy.setMaxReportDelay()`
- [ ] Add strategy to the keep3r job:
  - `keep3r_manager.addStrategy(strategy, 1_500_000, 1_500_000)`

>   - Tell Poolpi Tako `harvest` and `tend` gas usage. For instance 1.5m.

Reviewer 2: How do I find him?

Author: Removed. Added instructions to create a PR. Needs new Keep3r jobs deployed, TBD soon.
